node: cpumgr: metrics: add uncore cache alignment metrics

ffromani opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
Add alignment metrics for uncore (aka last-level) cache (KEP-4800)

#### Which issue(s) this PR fixes:
Part of the work needed for KEP 4800: https://github.com/kubernetes/enhancements/issues/4800

#### Special notes for your reviewer:
N/A

#### Does this PR introduce a user-facing change?
```release-note
Add metrics to track allocation of Uncore (aka last-level aka L3) Cache blocks
```

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/hold

mistakes in the e2e tests

ffromani commented:
/hold cancel

e2e tests should be fixed now

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
known issue, investigating

ffromani commented:
/hold

to ensure sufficient review

ffromani commented:
/hold cancel

needs more approval anyway :)
The PR is reviewable!

ffromani reviewed (commented):
notes to self:
things I'm ok with but not completely happy still

swatisehgal reviewed (commented):
Thanks for your work on this!
I left a few comments but will do another pass later. 

I found it a bit hard to review this because there was quiet some refactoring coupled with new additions. Can we try to split the work into smaller commits to split the refactoring work and new metrics addition?

ffromani committed:
node: cpumgr: test: abort on failure

abort the test in the unlikely case we fail to create
the Policy object. Exactly because this is unlikely
we should fail loudly.

In my case this happened because unrelated bad parameters,
which was hard to catch

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
node: cpumgr: test: use Equals for comparison

There's no need to use the generic reflect.DeepEqual,
we can use the cpuset Equals method in tests,
which is more idiomatic and a tad clearer.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
node: e2e: clarify test description

fix test name. Cosmetic only.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented on pkg/kubelet/cm/cpumanager/alignment.go:
```diff
@@ -0,0 +1,86 @@
+limitations under the License.
+*/
+
+package cpumanager
```
My only concern is that I want to merge the alignment logic with physical-cpus alignment logic once this PR merged and moving into `topology` could backfire on this.

ffromani commented on test/e2e_node/cpu_manager_metrics_test.go:
```diff
@@ -165,7 +181,7 @@ var _ = SIGDescribe("CPU Manager Metrics", framework.WithSerial(), feature.CPUMa
 		})
 
-		ginkgo.It("should return updated alignment counters when pod successfully run", func(ctx context.Context) {
+		ginkgo.It("should return updated SMT alignment counters when pod successfully run", func(ctx context.Context) {
```
right, split commits

ffromani commented on test/e2e_node/cpu_manager_metrics_test.go:
```diff
@@ -274,6 +290,88 @@ var _ = SIGDescribe("CPU Manager Metrics", framework.WithSerial(), feature.CPUMa
+
+			idFn := makeCustomPairID("scope", "boundary")
+			matchAlignmentMetrics := gstruct.MatchKeys(gstruct.IgnoreExtras, gstruct.Keys{
+				"kubelet_container_aligned_compute_resources_count": gstruct.MatchElements(idFn, gstruct.IgnoreExtras, gstruct.Elements{
+					"container::uncore_cache": timelessSample(1),
+				}),
+			})
```
Because AFAIK is more idiomatic in prometheus to have related logical metrics grouped in the same metric and differentiate by labels. There are other examples in the codebase, albeit our metrics are kinda inconsistent

ffromani committed:
node: cpumgr: metrics: add uncore cache alignment metrics

add missing metric about uncore / L3 / Last-Level cache alignment,
plus its e2e tests.

Exposing uncore alignment requires a bit of refactoring in the static
policy implementation because, differently from full PCPUs alignment
and NUMA alignment, can't be easily and safely inferred by construction.

The main reason for this is that uncore cache alignment is preferred,
not mandatory, thus the cpu allocator can legally use cross-uncore
allocation. Because of that, the final cpuset union step can
create a final cpuset which is not uncore-aligned even though all
its parts are uncore-aligned.

The safest way seems thus to run just a final uncore-alignment check
once the final cpuset is computed.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
/retest-required

unrelared apiserver failure

Tal-or commented on pkg/kubelet/cm/cpumanager/policy_static_test.go:
```diff
@@ -69,7 +69,10 @@ func (spt staticPolicyTest) PseudoClone() staticPolicyTest {
-	policy, _ := NewStaticPolicy(topoSingleSocketHT, 1, cpuset.New(), topologymanager.NewFakeManager(), nil)
+	policy, err := NewStaticPolicy(topoSingleSocketHT, 1, cpuset.New(), topologymanager.NewFakeManager(), nil)
+	if err != nil {
+		t.Fatalf("NewStaticP{olicy() failed: %v", err)
```
nit: a `{` slipped into the string

ffromani replied:
thanks, fixed

Tal-or commented on pkg/kubelet/cm/cpumanager/policy_static_test.go:
```diff
@@ -660,7 +660,7 @@ func runStaticPolicyTestCase(t *testing.T, testCase staticPolicyTest) {
 
-		if !reflect.DeepEqual(cset, testCase.expCSet) {
+		if !cset.Equals(testCase.expCSet) {
 			t.Errorf("StaticPolicy Allocate() error (%v). expected cpuset %v but got %v",
```
Maybe this out of the scope of this PR, but wouldn't it be to replace `%v` with `%s` for the `cpuset.CPUset` object and get nicer output?

ffromani replied:
why not? adding

swatisehgal commented:
/triage accepted
/priority important-longterm

ffromani committed:
node: cpumgr: use String representation in output

get nicer output for free

Signed-off-by: Francesco Romani <fromani@redhat.com>

ajcaldelas commented on pkg/kubelet/cm/cpumanager/topology/alignment.go:
```diff
@@ -0,0 +1,78 @@
+
+import (
+	"fmt"
+
```
would it make sense to add some logging here?

ffromani commented:
/retest-required

ffromani commented:
continuing on https://github.com/kubernetes/kubernetes/pull/130133
